src/System.Text.Json/src/System/Text/Json/JsonUtf8Reader.cs

```diff
@@ -0,0 +1,1274 @@
+                {
+                    int index = JsonConstants.EscapableChars.IndexOf(currentByte);
+                    if (index == -1)
+                        ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidCharacterWithinString, currentByte);
+
+                    if (currentByte == 'n')
+                    {
+                        // Escaped new line character
+                        _lineBytePosition = -1; // Should be 0, but we increment _lineBytePosition below already
+                        _lineNumber++;
+                    }
+                    else if (currentByte == 'u')
+                    {
+                        // Expecting 4 hex digits to follow the escaped 'u'
+                        _lineBytePosition++;
+                        int startIndex = i + 1;
+                        for (int j = startIndex; j < data.Length; j++)
+                        {
+                            byte nextByte = data[j];
+                            if ((uint)(nextByte - '0') > '9' - '0' && (uint)(nextByte - 'A') > 'F' - 'A' && (uint)(nextByte - 'a') > 'f' - 'a')
+                            {
+                                ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidCharacterWithinString, nextByte);
+                            }
+                            if (j - startIndex >= 4)
+                                break;
+                            _lineBytePosition++;
+                        }
+                        i += 4;
+                    }
+                    nextCharEscaped = false;
+                }
+                else if (currentByte < JsonConstants.Space)
+                    ThrowHelper.ThrowJsonReaderException(ref this, ExceptionResource.InvalidCharacterWithinString, currentByte);
+
+                _lineBytePosition++;
+            }
+        }
+
+        // https://tools.ietf.org/html/rfc7159#section-6
+        private bool TryGetNumber(ReadOnlySpan<byte> data, out int consumed)
+        {
+            // TODO: Optimize number processing and validation
+            // TODO: Cache the type of number into a private field (int, long, double, etc.)
+            Debug.Assert(data.Length > 0);
+
+            consumed = 0;
+            int i = 0;
+
+            ConsumeNumberResult signResult = ConsumeNegativeSign(ref data, ref i);
+            if (signResult == ConsumeNumberResult.NeedMoreData)
+                return false;
+
+            Debug.Assert(signResult == ConsumeNumberResult.OperationIncomplete);
+
+            byte nextByte = data[i];
+            Debug.Assert(nextByte >= '0' && nextByte <= '9');
+
+            if (nextByte == '0')
+            {
+                ConsumeNumberResult result = ConsumeZero(ref data, ref i);
+                if (result == ConsumeNumberResult.NeedMoreData)
+                    return false;
+                if (result == ConsumeNumberResult.Success)
+                    goto Done;
+
```

Comment:
Does it make sense to Debug.Assert the state(s) you think are proceeding?

Reply:
Hmm, I am not sure I follow. There are several potential states that can follow ConsumeZero based on the next input. What state are you referring to? The state of `ConsumeNumberResult`? If so, that is specific to the previous operation only (`ConsumeZero` in this case).

Reply:
I mean that you did an `if` for two of the three named states for `result`.  There's only one remaining defined state, which the code is assuming was returned.

So, `Debug.Assert(result == ConsumeNumberResult.OperationIncomplete);` is what I'm suggesting belongs here.